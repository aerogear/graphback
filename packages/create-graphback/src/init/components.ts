import { accessSync, mkdirSync } from 'fs';
import chalk from 'chalk';
import * as figlet from 'figlet';
import { prompt as ask } from 'inquirer';
import { logError, logInfo } from '../utils';
import { allTemplates, extractTemplate } from './starterTemplates';
import { Template } from './templateMetadata';

/**
 * Check if directory exists
 * @param path path of the directory
 * @param name name of the project folder
 */
function checkDirectory(path: string, name: string): void {
  try {
    accessSync(path);
    logError(
      `A folder with name ${name} exists. Remove it or try another name.`
    );
    process.exit(0);
  } catch (error) {
    return;
  }
}

/**
 * choose a template from available templates
 */
async function chooseTemplate(filter: string = ''): Promise<Template> {
  const regex = new RegExp(`.*${filter}.*`, 'i');
  const displayedTemplates = allTemplates.filter((template: Template) =>
    regex.test(`${chalk.green(template.name)} ${template.description}`)
  );
  if (!displayedTemplates.length) {
    logInfo(`create-graphback could not find templates matching the given filter: "${filter}".
You can either change the given filter or not pass the option to display the full list.`);
    process.exit(0);
  }

  logInfo(`Graphback init can create your app from following templates:
  ${displayedTemplates
    .map((template: Template) => {
      return `\n${chalk.green(template.name)}: \n${template.description}`;
    })
    .join('\n')}
  `);
  const { templateName } = await ask([
    {
      type: 'list',
      name: 'templateName',
      message: 'Choose a template to bootstrap',
      choices: displayedTemplates.map((t: Template) => t.name)
    }
  ]);

  return displayedTemplates.find((t: Template) => t.name === templateName);
}

/**
 * check if template name is valid or not
 * @param templateName name of the template provided
 */
function checkTemplateName(templateName: string): void {
  const availableTemplates = allTemplates.map((t: Template) => t.name);
  if (availableTemplates.includes(templateName)) {
    return;
  }
  logError(
    "Template with given name doesn't exist. Give one of available ones or simply choose by not providing a template name"
  );
  process.exit(0);
}

/**
 * assign template details from the given input or choice
 * @param templateName name of the template provided(if any)
 */
async function assignTemplate(
  templateName: string,
  filter: string
): Promise<Template> {
  let template;
  if (templateName) {
    checkTemplateName(templateName);
    template = allTemplates.find((t: Template) => t.name === templateName);
  } else {
    template = await chooseTemplate(filter);
  }

  return template;
}

function postSetupMessage(name: string): string {
  return `
GraphQL server successfully bootstrapped :rocket:

Next Steps:
1. Change directory into project folder - ${chalk.cyan(`cd ${name}`)}
2. Follow template README.md to start server
`;
}

/**
 * Build template from user provided url
 */
function buildTemplateFromGithub(templateUrl: string) {
  const url = templateUrl.split('#');

  return {
    name: 'Users Github template',
    description: 'User provided template',
    repos: [
      {
        uri: url[0],
        branch: url[1] || 'master',
        path: '/'
      }
    ]
  };
}

/**
 * init command handler
 * @param name name of project folder
 * @param templateName name of the template provided(if any)
 * @param templateUrl github url to the template
 */
export async function init(
  name: string,
  templateName?: string,
  templateUrl?: string,
  filter?: string
) {
  logInfo(
    chalk.yellow(figlet.textSync('Graphback', { horizontalLayout: 'full' }))
  );
  const path: string = `${process.cwd()}/${name}`;
  checkDirectory(path, name);
  let template: Template;
  if (templateUrl) {
    template = buildTemplateFromGithub(templateUrl);
  } else {
    template = await assignTemplate(templateName, filter);
  }
  mkdirSync(path);
  logInfo(`
Bootstraping graphql server :dizzy: :sparkles:`);
  await extractTemplate(template, name);
  process.chdir(name);
  logInfo(postSetupMessage(name));
}
