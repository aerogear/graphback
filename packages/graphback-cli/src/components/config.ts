import chalk from 'chalk';
import * as execa from 'execa'
import * as figlet from 'figlet'
import { askForClient, chooseDatabase, createConfig } from '../templates/configTemplates'
import { addModel, createModel, ModelTemplate } from '../templates/modelTemplates';
import { logInfo } from '../utils';

/**
 * Install dependencies, currently only npm
 */
export async function installDependencies(database: string): Promise<void> {
  logInfo('Installing dependencies')
  await execa('npm', ['i'])
  if (database === 'pg') {
    await execa('npm', ['i', '-S', 'pg'])
  } else if (database === 'sqlite3') {
    await execa('npm', ['i', '-S', 'sqlite3'])
  }
  //Using opiniated layer
  await execa('npm', ['i', '-S', '@graphback/runtime-knex'])
}

function postSetupMessage(): string {
  return `
Graphback configuration successfully bootstrapped :rocket:

Next Steps:
1. Review your configuration in "graphqlrc.yaml" file
2. Edit the *.graphql files inside your model folder.
3. Run  generate command to generate your schema and resolvers
`
}

/**
 * initConfig options
 */
export interface InitOptions {
  client?: boolean,
  //Type of database
  database?: string
  //example model
  model?: ModelTemplate
  //skip install
  skipInstall?: boolean
  // Skip config generation
  skipConfig?: boolean
}

/**
 * config command handler
 * @param options used for initializting config without asking user for input
 */
export async function initConfig(options: InitOptions = {}) {
  logInfo(chalk.yellow(
    figlet.textSync('Graphback', { horizontalLayout: 'full' })
  ));

  const { modelName, content } = options.model || await createModel()
  const database = options.database || await chooseDatabase()
  const client = options.client || await askForClient()
  addModel("", modelName, content)
  if (!options.skipInstall) {
    await installDependencies(database)
  }
  if (!options.skipConfig) {
    await createConfig(database, client)
  }
  logInfo(postSetupMessage())
}
