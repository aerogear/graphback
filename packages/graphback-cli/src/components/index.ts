export * from './generate'
export * from './config'
