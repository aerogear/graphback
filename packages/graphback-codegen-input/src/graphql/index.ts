
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"