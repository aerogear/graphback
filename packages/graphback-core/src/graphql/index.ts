
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"
export * from "./directives"