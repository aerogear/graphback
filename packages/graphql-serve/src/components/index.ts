export * from './serveHandler';
export * from './printSchemaHandler';