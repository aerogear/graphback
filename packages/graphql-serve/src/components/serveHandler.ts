import { buildGraphbackServer } from "../GraphbackServer";
import { getGraphbackServerConfig } from "../GraphbackServerConfig";

export const serveHandler = async (argv: { model?: string, port?: number }): Promise<void> => {
    const graphbackServerConfig = await getGraphbackServerConfig({ modeldir: argv.model });
    const server = await buildGraphbackServer(graphbackServerConfig);
    if ('port' in argv) {
        const portNumber = argv.port;
        if (isNaN(portNumber)) {
            console.log("\nSpecified port number is NaN, terminating...\n");
        } else {
            console.log("\nStarting server...\n");
            server.start(portNumber);
        }
    } else {
        console.log("\nNo port number specified.\nStarting server on random available port...\n");
        server.start();
    }
};
