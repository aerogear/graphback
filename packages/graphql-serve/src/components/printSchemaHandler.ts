import { GraphbackRuntime } from 'graphback'
import { printSchema } from 'graphql';
import { getGraphbackServerConfig } from "../GraphbackServerConfig";
import { loadSchema }  from '../loadSchema';

export const printSchemaHandler = async (argv: {model?: string}) => {
    const graphbackServerConfig = await getGraphbackServerConfig({ modeldir: argv.model });

    const graphbackConfig = graphbackServerConfig.graphback;
    const schemaText = loadSchema(graphbackConfig.model);

    const runtimeEngine = new GraphbackRuntime(schemaText, graphbackConfig);

    const schema = runtimeEngine.getMetadata().getSchema();

    console.log("Generated schema:\n");
    console.log(printSchema(schema));
}