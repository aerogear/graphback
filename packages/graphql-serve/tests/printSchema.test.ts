import { printSchemaHandler } from '../src/components/printSchemaHandler';
import { buildSchema } from 'graphql';

export const expectedUserSchema = `"""
Directs the executor to skip this field or fragment when the \`if\` argument is true.
"""
directive @skip(
  """Skipped when true."""
  if: Boolean!
) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"""
Directs the executor to include this field or fragment only when the \`if\` argument is true.
"""
directive @include(
  """Included when true."""
  if: Boolean!
) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"""Marks an element of a GraphQL schema as no longer supported."""
directive @deprecated(
  """
  Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).
  """
  reason: String = "No longer supported"
) on FIELD_DEFINITION | ENUM_VALUE

"""Exposes a URL that specifies the behaviour of this scalar."""
directive @specifiedBy(
  """The URL that specifies the behaviour of this scalar."""
  url: String!
) on SCALAR

input CreateUserInput {
  id: ID
  name: String
}

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID]
}

input MutateUserInput {
  id: ID!
  name: String
}

type Mutation {
  createUser(input: CreateUserInput!): User!
  updateUser(input: MutateUserInput!): User!
  deleteUser(input: MutateUserInput!): User!
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getUser(id: ID!): User
  findUsers(filter: UserFilter, page: PageRequest, orderBy: OrderByInput): UserResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newUser(filter: UserSubscriptionFilter): User!
  updatedUser(filter: UserSubscriptionFilter): User!
  deletedUser(filter: UserSubscriptionFilter): User!
}

"""@model"""
type User {
  id: ID!
  name: String
}

input UserFilter {
  id: IDInput
  name: StringInput
  and: [UserFilter]
  or: [UserFilter]
  not: UserFilter
}

type UserResultList {
  items: [User]!
  offset: Int
  limit: Int
  count: Int
}

input UserSubscriptionFilter {
  id: ID
  name: String
}
`

beforeEach(() => process.chdir(__dirname))

test('printSchema from GraphQL file', async () => {
  const modelFile = './files/user-model.graphql'

  const schemaSDL = await printSchemaHandler({ model: modelFile })

  expect(schemaSDL).toBeDefined()
  expect(schemaSDL).toMatchSnapshot()
  expect(buildSchema(schemaSDL).astNode).toEqual(buildSchema(expectedUserSchema).astNode)
})

test('printSchema from directory', async () => {
  const modelFile = './files'

  const schemaSDL = await printSchemaHandler({ model: modelFile })

  expect(schemaSDL).toBeDefined()
  expect(buildSchema(schemaSDL).astNode).toEqual(buildSchema(expectedUserSchema).astNode)
})
