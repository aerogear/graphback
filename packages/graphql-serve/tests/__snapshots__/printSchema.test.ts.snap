// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printSchema from GraphQL file 1`] = `
"\\"\\"\\"Exposes a URL that specifies the behaviour of this scalar.\\"\\"\\"
directive @specifiedBy(
  \\"\\"\\"The URL that specifies the behaviour of this scalar.\\"\\"\\"
  url: String!
) on SCALAR

input CreateUserInput {
  name: String
}

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID!]
}

input MutateUserInput {
  id: ID!
  name: String
}

type Mutation {
  createUser(input: CreateUserInput!): User
  updateUser(input: MutateUserInput!): User
  deleteUser(input: MutateUserInput!): User
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getUser(id: ID!): User
  findUsers(filter: UserFilter, page: PageRequest, orderBy: OrderByInput): UserResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newUser(filter: UserSubscriptionFilter): User!
  updatedUser(filter: UserSubscriptionFilter): User!
  deletedUser(filter: UserSubscriptionFilter): User!
}

\\"\\"\\"@model\\"\\"\\"
type User {
  id: ID!
  name: String
}

input UserFilter {
  id: IDInput
  name: StringInput
  and: [UserFilter!]
  or: [UserFilter!]
  not: UserFilter
}

type UserResultList {
  items: [User]!
  offset: Int
  limit: Int
  count: Int
}

input UserSubscriptionFilter {
  and: [UserSubscriptionFilter!]
  or: [UserSubscriptionFilter!]
  not: UserSubscriptionFilter
  id: IDInput
  name: StringInput
}"
`;
